device/thunder_ripple_sdk/src/client/plugin_manager.rs

```diff
@@ -272,39 +307,53 @@ impl PluginManager {
         let state = match self.plugin_states.get(&callsign) {
             Some(state) => state.clone(),
             None => {
-                // No state is know, go fetch the state from thunder controller and store in cache
+                // No state is known, go fetch the state from thunder controller and store in cache
                 let state = self.current_plugin_state(callsign.clone()).await;
                 self.plugin_states.insert(callsign.clone(), state.clone());
                 state
             }
         };
-        if state.is_activated() {
-            return PluginActivatedResult::Ready;
-        }
-        // If plugin is not activated, then add a subscriber for state change and then activate
-        let (sub_tx, sub_rx) = oneshot::channel::<()>();
 
-        self.state_subscribers.push(ActivationSubscriber {
-            callsign: callsign.clone(),
-            callback: sub_tx,
-        });
-        if trigger_activation && self.activate_plugin(callsign).await.message.is_null() {
-            return PluginActivatedResult::Error;
+        if state.is_missing() {
+            PluginActivatedResult::Error
+        } else if state.is_activated() {
+            PluginActivatedResult::Ready
+        } else {
+            // means plugin needs activation and any call needs to wait for the activation to be successful before proceeding
+            // If plugin is not activated, then add a subscriber for state change and then activate
+            let (sub_tx, sub_rx) = oneshot::channel::<PluginState>();
+
+            self.state_subscribers.push(ActivationSubscriber {
+                callsign: callsign.clone(),
+                callback: sub_tx,
+            });
+            if trigger_activation {
+                return match self.activate_plugin(callsign).await {
+                    PluginState::Activated => PluginActivatedResult::Ready,
+                    PluginState::Missing => PluginActivatedResult::Error,
+                    _ => PluginActivatedResult::Pending(sub_rx),
+                };
+            }
+            PluginActivatedResult::Pending(sub_rx)
         }
-        PluginActivatedResult::Pending(sub_rx)
     }
 
-    async fn activate_plugin(&self, callsign: String) -> DeviceResponseMessage {
+    async fn activate_plugin(&self, callsign: String) -> PluginState {
         let r = ThunderActivatePluginParams { callsign };
-        self.thunder_client
+        let resp = self
+            .thunder_client
             .clone()
             .call(DeviceCallRequest {
                 method: Controller.method("activate"),
                 params: Some(DeviceChannelParams::Json(
                     serde_json::to_string(&r).unwrap(),
                 )),
             })
-            .await
+            .await;
+        if let Ok(plugin_error) = serde_json::from_value::<ThunderError>(resp.message) {
```

Comment:
I couldn’t locate the code where we are responding DeviceCallRequest with ThunderError in resp.message. 
Are we setting `Value::Null` to resp.message on Thunder error ?

Reply:
We create a custom internal error for the case where we didnt send the request to Thunder
```
json!({"error": "pre send error"}
```